### PR TITLE
[TorchAcc] Integrate TorchAcc and provide a sft example of qwen-72b-chat.

### PR DESCRIPTION
**Pull request by baoleai**
TorchAcc is a training acceleration engine for LLMs developed by Alibaba's PAI team. It is based on compilation optimization methods and provides support for various LLM models. TorchAcc offers common distributed strategies such as FSDP, TP+SP+PP, and ZeRO. Currently, TorchAcc is only available internally. This PR provides an example of using TorchAcc to accelerate swift SFT lora on Qwen-72B-Chat with 2*A800.

### CODE REVIEW

File: swift/trainers/trainers.py [hunk 1]
```diff
@@ -179,24 +179,45 @@ def compute_loss(self, model, inputs, return_outputs=None):
...
-            if 'acc' not in self._custom_metrics:
-                self._custom_metrics['acc'] = torch.tensor(0.).to(
-                    self.args.device)
-            self._custom_metrics[
-                'acc'] += acc / self.args.gradient_accumulation_steps
+            acc_strategy = getattr(self.args, 'acc_strategy', 'token')
+            acc: Tensor
+            if acc_strategy == 'sentence':
+                acc_list = []
+                for i, m in enumerate(masks):
```
Comment (Jintao-Huang) [inline swift/trainers/trainers.py]: masks not defined

File: swift/trainers/trainers.py [hunk 2]
```diff
@@ -179,24 +179,45 @@ def compute_loss(self, model, inputs, return_outputs=None):
...
+                    acc_list.append(
+                        torch.all(preds[i, m] == labels[i, m]).to(
+                            torch.int64).item())
+                acc = torch.tensor(
+                    acc_list, device=preds.device).float().mean()
+            else:
+                acc = (preds == labels).float().mean()
+            if model.training:
```
Comment (Jintao-Huang) [inline swift/trainers/trainers.py]: The code here has a bit high redundancy. Is it possible to optimize it?

File: swift/llm/utils/utils.py [hunk 3]
```diff
@@ -724,6 +756,8 @@ def get_time_info(log_history: List[Dict[str, Any]],
         _old_ddp_init(self, model, *args, **kwargs))
     transformers.modeling_utils.get_balanced_memory = lambda *args, **kwargs: None
     transformers.modeling_utils.infer_auto_device_map = _infer_auto_device_map_patch
+
+if is_ddp_plus_mp() or use_torchacc():
```
Comment (Jintao-Huang) [inline swift/llm/utils/utils.py:763]: Is torchacc compatible with mp+ddp?

File: swift/llm/utils/model.py [hunk 4]
```diff
@@ -1641,7 +1641,7 @@ def get_model_tokenizer(
     ignore_file_pattern = model_info['ignore_file_pattern']
     if model_kwargs is None:
         model_kwargs = {}
-    if 'device_map' not in model_kwargs:
+    if (not use_torchacc()) and ('device_map' not in model_kwargs):
```
Comment (Jintao-Huang) [inline swift/llm/utils/model.py]: 'device_map' not in model_kwargs and not use_torchacc()

File: swift/llm/sft.py [hunk 5]
```diff
@@ -199,7 +234,8 @@ def llm_sft(args: SftArguments) -> Dict[str, Union[str, Any]]:
         push_to_hub=args.push_to_hub,
         resume_from_checkpoint=args.resume_from_checkpoint,
         ddp_backend=args.ddp_backend,
-        gradient_checkpointing=args.gradient_checkpointing,
+        gradient_checkpointing=False
+        if use_torchacc() else args.gradient_checkpointing,
```
Comment (Jintao-Huang) [inline swift/llm/sft.py]: Write the modifications for gradient_checkpointing in the argument.py directly.

File: swift/llm/sft.py [hunk 6]
```diff
@@ -79,6 +90,27 @@ def llm_sft(args: SftArguments) -> Dict[str, Union[str, Any]]:
     logger.info(model_info)
     logger.info(model)
 
+    if use_torchacc():
+        import torchacc as ta
+
+        def get_ta_config():
```
Comment (Jintao-Huang) [inline swift/llm/sft.py]: Please don't write this function in the main, may be utils

File: swift/llm/sft.py [hunk 7]
```diff
@@ -79,6 +90,27 @@ def llm_sft(args: SftArguments) -> Dict[str, Union[str, Any]]:
...
+            config = ta.Config()
+            config.compute.fp16 = False
+            config.compute.bf16 = True
+
+            config.memory.gc = True
+            if config.memory.gc:
+                config.memory.gc_cls = {'QWenBlock'}
```
Comment (Jintao-Huang) [inline swift/llm/sft.py]: why only qwen

File: swift/trainers/arguments.py [hunk 8]
```diff
@@ -52,4 +52,7 @@ class TrainingArguments(SwiftArgumentsMixin, HfTrainingArguments):
 @dataclass
 class Seq2SeqTrainingArguments(SwiftArgumentsMixin,
                                HfSeq2SeqTrainingArguments):
-    pass
+
+    @property
+    def place_model_on_device(self):
+        return False if use_torchacc() else super().place_model_on_device
```
Comment (Jintao-Huang) [inline swift/trainers/arguments.py:58]: Does torchacc only support LLM? Can it be written in SwiftArgumentsMixin

File: examples/pytorch/llm/scripts/qwen_72b_chat/torchacc/lora_fsdp_sft.sh [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+	--dataset codefuse-python-en \
+	--sft_type lora \
+  --output_dir output \
+  --num_train_epochs 1 \
+  --max_length 1024 \
+  --batch_size 1 \
+  --use_flash_attn true \
+  --gradient_accumulation_steps 1 \
+  --tuner_backend 'peft' \
+  --eval_steps 2000000 \
```
Comment (Jintao-Huang) [inline examples/pytorch/llm/scripts/qwen_72b_chat/torchacc/lora_fsdp_sft.sh:25]: Why so large

File: swift/llm/sft.py [hunk 10]
```diff
@@ -254,6 +290,9 @@ def llm_sft(args: SftArguments) -> Dict[str, Union[str, Any]]:
         callbacks=callbacks,
         **trainer_kwargs)
     trainer.sft_args = args
+    if use_torchacc():
+        trainer.label_names = label_names
```
Comment (Jintao-Huang) [inline swift/llm/sft.py:288]: Can you explain this part? I didn't understand it.

File: swift/llm/sft.py [hunk 11]
```diff
@@ -215,7 +251,7 @@ def llm_sft(args: SftArguments) -> Dict[str, Union[str, Any]]:
         logging_first_step=True,
         **kwargs)
 
-    if args.gradient_checkpointing:
+    if args.gradient_checkpointing and not use_torchacc():
```
Comment (Jintao-Huang) [inline swift/llm/sft.py]: Write the modifications for gradient_checkpointing in the argument.py directly.

File: examples/pytorch/llm/scripts/qwen_72b_chat/torchacc/lora_fsdp_sft.sh [hunk 12]
```diff
@@ -0,0 +1,27 @@
+export USE_TORCHACC=1
+export XLA_FLAGS='--xla_gpu_enable_latency_hiding_scheduler=true --xla_enable_pt_all_reduce_optimizer=false --xla_gpu_enable_async_all_gather=true --xla_gpu_enable_async_reduce_scatter=true --xla_multiheap_size_constraint_per_heap=4294967296 --xla_disable_hlo_passes=rematerialization,all-gather-combiner,all-reduce-combiner,reduce-scatter-combiner'
+export ACC_FLASH_ATTN=1
+export LOW_CPU_MEM_USAGE=1
+export XLA_IR_SHAPE_CACHE_SIZE=100000000
+export XLA_ALLOCATOR_FRACTION=0.97
+
+MASTER_ADDR=127.0.0.1 \
+MASTER_PORT=9010 \
+NPROC_PER_NODE=2 \
```
Comment (Jintao-Huang) [inline examples/pytorch/llm/scripts/qwen_72b_chat/torchacc/lora_fsdp_sft.sh:10]: Can you provide an example that demonstrates the advantages of torchacc? and create a related introductory document in LLM to attract users to use it.

File: swift/llm/utils/utils.py [hunk 13]
```diff
@@ -298,15 +300,35 @@ def stat_dataset(llm_dataset: Dataset) -> str:
...
 
 
-def data_collate_fn(batch: List[Dict[str, Any]],
-                    tokenizer: PreTrainedTokenizerBase,
-                    padding_to: Optional[int] = None) -> Dict[str, Any]:
+def get_bucket_sizes(max_length: int) -> List[int]:
+    return [max_length // 4 * (i + 1) for i in range(4)]
+
+
+def _get_bucket(bucket_sizes, data_length):
```
Comment (Jintao-Huang) [inline swift/llm/utils/utils.py:307]: Please add annotations.

File: swift/llm/utils/utils.py [hunk 14]
```diff
@@ -332,6 +354,16 @@ def data_collate_fn(batch: List[Dict[str, Any]],
         attention_mask, batch_first=True, padding_value=0)
     labels = pad_sequence(labels, batch_first=True, padding_value=-100)
 
+    if padding_to is None and use_torchacc():
```
Comment (Jintao-Huang) [inline swift/llm/utils/utils.py:360]: Can I ask if padding_to and bucket_sizes cannot be set simultaneously

**Comment by Jintao-Huang** [comment]
Thank you for your pull request. I have some questions and have provided some comments. Please take a look and let me know if my requirements are reasonable. Thank you.

File: examples/pytorch/llm/scripts/qwen_72b_chat/torchacc/lora_fsdp_sft.sh [hunk 15]
```diff
@@ -0,0 +1,27 @@
+export USE_TORCHACC=1
+export XLA_FLAGS='--xla_gpu_enable_latency_hiding_scheduler=true --xla_enable_pt_all_reduce_optimizer=false --xla_gpu_enable_async_all_gather=true --xla_gpu_enable_async_reduce_scatter=true --xla_multiheap_size_constraint_per_heap=4294967296 --xla_disable_hlo_passes=rematerialization,all-gather-combiner,all-reduce-combiner,reduce-scatter-combiner'
```
Comment (anw90) [inline examples/pytorch/llm/scripts/qwen_72b_chat/torchacc/lora_fsdp_sft.sh]: Maybe we should move the ENV variables from Line2 to Line6 to TorchAcc.

File: swift/llm/utils/utils.py (hunk 3 above)
Comment (baoleai) [reply swift/llm/utils/utils.py:763]: No. But TorchAcc provides different parallel strategies itself, including MP + DP.

File: examples/pytorch/llm/scripts/qwen_72b_chat/torchacc/lora_fsdp_sft.sh (hunk 9 above)
Comment (baoleai) [reply examples/pytorch/llm/scripts/qwen_72b_chat/torchacc/lora_fsdp_sft.sh:25]: Currently, there is a performance issue in evaluation, I will fix it assp.

File: swift/llm/utils/utils.py (hunk 14 above)
Comment (baoleai) [reply swift/llm/utils/utils.py:360]: Yes, bucketing of sequence lengths is only needed when sequence length is dynamic i.e. padding_to is None.

File: swift/trainers/arguments.py (hunk 8 above)
Comment (baoleai) [reply swift/trainers/arguments.py:58]: Yes. currently torchacc only supports LLM.

File: swift/trainers/trainers.py (hunk 1 above)
Comment (baoleai) [reply swift/trainers/trainers.py]: fixed

File: swift/trainers/trainers.py (hunk 2 above)
Comment (baoleai) [reply swift/trainers/trainers.py]: fixed

File: swift/llm/utils/model.py (hunk 4 above)
Comment (baoleai) [reply swift/llm/utils/model.py]: fixed

File: swift/llm/sft.py (hunk 5 above)
Comment (baoleai) [reply swift/llm/sft.py]: fixed

File: swift/llm/sft.py (hunk 6 above)
Comment (baoleai) [reply swift/llm/sft.py]: move it to accelerator.py

File: swift/llm/sft.py (hunk 7 above)
Comment (baoleai) [reply swift/llm/sft.py]: replace by argument.

File: swift/llm/sft.py (hunk 10 above)
Comment (baoleai) [reply swift/llm/sft.py:288]: added.

File: swift/llm/sft.py (hunk 11 above)
Comment (baoleai) [reply swift/llm/sft.py]: done

File: swift/llm/utils/utils.py (hunk 13 above)
Comment (baoleai) [reply swift/llm/utils/utils.py:307]: added

File: examples/pytorch/llm/scripts/qwen_72b_chat/torchacc/lora_fsdp_sft.sh (hunk 15 above)
Comment (baoleai) [reply examples/pytorch/llm/scripts/qwen_72b_chat/torchacc/lora_fsdp_sft.sh]: Removed certain flags that are pre-configured within TorchAcc.

**Comment by Xu-Chen** [comment]
With this PR, can we fully fine-tune the Qwen-72b model on a single machine 8*A800 or multinode 4*8*A800?

**Comment by baoleai** [comment]
> With this PR, can we fully fine-tune the Qwen-72b model on a single machine 8 x A800 or multinode 4 x 8 x A800?

Although this PR does not provide an example of a fully sft, fully QWen-72B sft can be run on a 4 x 8 x A800 configuration. You need to modify the script and change the sft_type to 'full', and our subsequent PR will provide a complete example.

**Comment by Xu-Chen** [comment]
> > With this PR, can we fully fine-tune the Qwen-72b model on a single machine 8 x A800 or multinode 4 x 8 x A800?
> 
> Although this PR does not provide an example of a fully sft, fully QWen-72B sft can be run on a 4 x 8 x A800 configuration. You need to modify the script and change the sft_type to 'full', and our subsequent PR will provide a complete example.

Thank you

File: examples/pytorch/llm/scripts/qwen_72b_chat/torchacc/full_fsdp_sft.sh [hunk 16]
```diff
@@ -0,0 +1,34 @@
+# Experimental environment: 4 * 8*A100
+# 80GB GPU memory
+# Note: TorchAcc is currently only available internally.
```
Comment (Xu-Chen) [inline examples/pytorch/llm/scripts/qwen_72b_chat/torchacc/full_fsdp_sft.sh:3]: @baoleai when can we use TorchAcc?

Comment (baoleai) [reply examples/pytorch/llm/scripts/qwen_72b_chat/torchacc/full_fsdp_sft.sh:3]: Thank you for your interest. We are currently undergoing comprehensive testing and optimization in multiple scenarios. Once testing is complete, we will release a customized runtime environment for TorchAcc with detailed deployment and usage guidelines. Ultimately, we will open source TorchAcc with rigorous testing and clean code.
